Editorial review: Document the Captured Surface Control API (#39982)
mdn/content: files/en-us/web/api/screen_capture_api/captured_surface_control/index.md

The `updateZoomButtonState()` function avoids this issue by first making sure both the "Zoom out" and "Zoom in" buttons are enabled. It then does two checks:

- If the current zoom level (returned by the `CaptureController.zoomLevel` property) is equal to the minimum supported zoom level (stored in the first value of the `zoomLevels` array), we disable the "Zoom out" button so the user can't zoom out any further.
- If the current zoom level is equal to the maximum supported zoom level (stored in the last value of the `zoomLevels` array), we disable the "Zoom in" button so the user can't zoom in any further.

```js live-sample___surface-control-demo
function updateZoomButtonState() {
  decBtn.disabled = false;
  incBtn.disabled = false;
  if (controller.zoomLevel === zoomLevels[0]) {
    decBtn.disabled = true;
  } else if (controller.zoomLevel === zoomLevels[zoomLevels.length - 1]) {
    incBtn.disabled = true;
  }
}
```

## Finished demo

The finished demo renders like this:

{{EmbedLiveSample("surface-control-demo", , "500px", , , , "captured-surface-control")}}

> EDITORIAL: The demo works, but when you scroll the captured display surface, it crashes the browser tab.

Reviewer: Really?! Have you filed a Chromium bug? (If so, please assign it to me.)

Author: Done: https://issues.chromium.org/issues/427295493.

I haven't assigned it to you because I don't think I have the permissions to do so. Mentioning @guidou so that it is handled while you are OOO.

The Captured Surface Control API tackles these problems by allowing app developers to implement a limited set of features that can be used by conference participants to control the captured display surface from directly within the app, without compromising security.

Currently these are:

1. Zooming the captured display surface.
2. Using mouse wheel/touchpad gestures (and other equivalents) to scroll the captured display surface.

This functionality is all accessed via the {{domxref("CaptureController")}} object. To control a captured display surface, a capture controller must be passed into a {{domxref("MediaDevices.getDisplayMedia()")}} call inside its options object:

```js
controller = new CaptureController();

const displayMediaOptions = {
  controller,
};

videoElem.srcObject =
  await navigator.mediaDevices.getDisplayMedia(displayMediaOptions);
```

The controller can then be used to, for example, zoom the captured display surface in:

```js
controller.increaseZoomLevel();
```

In this article we will walk through the code for a basic screen sharing app that shows how to implement such features.

> ## A note on permissions
> 
> Usage of the Captured Surface Control API can be controlled via the {{HTTPHeader("Permissions-Policy")}} {{HTTPHeader("Permissions-Policy/captured-surface-control", "captured-surface-control")}} directive, or the equivalent {{HTMLElement("iframe")}} [`allow`](/en-US/docs/Web/HTML/Reference/Elements/iframe#allow)) attribute value:
> 
> ```html
> <iframe allow="captured-surface-control" src="/some-other-document.html"
>   >…</iframe
> >
> ```
> 
> Specifically, the {{domxref("CaptureController.forwardWheel", "forwardWheel()")}}, {{domxref("CaptureController.increaseZoomLevel", "increaseZoomLevel()")}}, {{domxref("CaptureController.decreaseZoomLevel", "decreaseZoomLevel()")}}, and {{domxref("CaptureController.resetZoomLevel", "resetZoomLevel()")}} methods are controlled by this directive.
> 
> The default allowlist for `captured-surface-control` is `self`, which lets any content within the same origin use Captured Surface Control.
> 
> ## App HTML

Reviewer 2: We should also say a few words about the broader security considerations here. You might want to link to the other section. Here's a quick suggestion.

```suggestion
## A note on permissions

A website can control access to the Captured Surface Control API using the {{HTTPHeader("Permissions-Policy")}} {{HTTPHeader("Permissions-Policy/captured-surface-control", "captured-surface-control")}} directive, or the equivalent {{HTMLElement("iframe")}} [`allow`](/en-US/docs/Web/HTML/Reference/Elements/iframe#allow)) attribute value:

```html
<iframe allow="captured-surface-control" src="/some-other-document.html"
  >…</iframe
>
```

Specifically, the {{domxref("CaptureController.forwardWheel", "forwardWheel()")}}, {{domxref("CaptureController.increaseZoomLevel", "increaseZoomLevel()")}}, {{domxref("CaptureController.decreaseZoomLevel", "decreaseZoomLevel()")}}, and {{domxref("CaptureController.resetZoomLevel", "resetZoomLevel()")}} methods are controlled by this directive.

The default allowlist for `captured-surface-control` is `self`, which lets any content within the same origin use Captured Surface Control.

In addition, if permission is granted by the policy, the user must also grant permission.
This can either be explicit permission, granted by responding to a prompt, or implicit permission granted by interacting with a control that calls one of the methods (transient activation) when user permission has not been explicitly denied.

## App HTML
```

Author: Your wording largely sounded good here. I've mostly used it, with some tweaks. I've also added a link, as suggested.

```js
controller.increaseZoomLevel();
```

In this article we will walk through the code for a basic screen sharing app that shows how to implement such features.

## A note on permissions

A website can control access to the Captured Surface Control API using the {{HTTPHeader("Permissions-Policy")}} {{HTTPHeader("Permissions-Policy/captured-surface-control", "captured-surface-control")}} directive, or the equivalent {{HTMLElement("iframe")}} [`allow`](/en-US/docs/Web/HTML/Reference/Elements/iframe#allow) attribute value:

```html
<iframe allow="captured-surface-control" src="/some-other-document.html">
  ...
</iframe>
```

Specifically, the {{domxref("CaptureController.forwardWheel", "forwardWheel()")}}, {{domxref("CaptureController.increaseZoomLevel", "increaseZoomLevel()")}}, {{domxref("CaptureController.decreaseZoomLevel", "decreaseZoomLevel()")}}, and {{domxref("CaptureController.resetZoomLevel", "resetZoomLevel()")}} methods are controlled by this directive.

The default allowlist for `captured-surface-control` is `self`, which lets any content within the same origin use Captured Surface Control.

> In addition, if permission is granted by the policy, the user must also grant permission. This can either be explicit permission, granted by responding to a prompt, or implicit permission granted by interacting with a control that calls one of the methods ([transient activation](/en-US/docs/Glossary/Transient_activation)) when user permission has not been explicitly denied.

Reviewer 2: A bit of a nit, but I wanted to make it clear that the policy gates the ability of a user to offer permission

```suggestion
If permission is allowed by the website policy, the user may then grant permission (or deny) access to the controlled APIs. This can either be explicit permission, granted by responding to a prompt, or implicit permission granted by interacting with a control that calls one of the methods ([transient activation](/en-US/docs/Glossary/Transient_activation)) when user permission has not been explicitly denied.
```
